File: tools/convert.py
# convert.py - Simple script to convert YAML to EDS
import sys
import os
import yaml
import copy
from eds_codec import json_to_eds

SCHEMA_VERSION = "1.0.0"

# Load defaults from schema.yaml
def load_schema():
    """Load element type defaults from schema.yaml."""
    global SCHEMA_VERSION
    schema_path = os.path.join(os.path.dirname(__file__), "schema.yaml")
    with open(schema_path, "r", encoding="utf-8") as f:
        schema = yaml.safe_load(f)

    # Get schema version
    SCHEMA_VERSION = schema.get("version", "1.0.0")

    defaults = {}
    for type_name, type_def in schema.get("types", {}).items():
        defaults[type_name] = type_def.get("defaults", {})
    return defaults

DEFAULTS = load_schema()

def normalize_node(node):
    """Convert shorthand syntax to standard format.

    Supports:
    - "Lichtcircuit" -> {type: Lichtcircuit}
    - Lichtcircuit: {props: ...} -> {type: Lichtcircuit, props: ...}
    - {type: Lichtcircuit, ...} -> unchanged
    """
    if isinstance(node, str):
        # Simple string: "Lichtcircuit"
        return {"type": node}

    if isinstance(node, dict):
        if "type" in node:
            # Already in standard format
            return node

        # Check if it's a type-name-as-key format
        for type_name in DEFAULTS.keys():
            if type_name in node:
                # Format: Lichtcircuit: {props: ..., children: ...}
                # Or:     Lichtcircuit: {type_lamp: spot, children: ...}
                value = node[type_name] or {}
                result = {"type": type_name}
                if isinstance(value, dict):
                    # Separate children from props
                    children = value.pop("children", None)
                    props = value.pop("props", None)
                    # Remaining keys are direct props
                    if value:
                        props = {**(props or {}), **value}
                    if props:
                        result["props"] = props
                    if children:
                        result["children"] = children
                return result

    return node

def apply_defaults(node_type, props):
    """Apply defaults for a node type, then overlay with provided props."""
    defaults = DEFAULTS.get(node_type, {})
    result = copy.deepcopy(defaults)
    result.update(props)
    return result

def compile_tree(tree, properties=None, sitplan=None, yaml_dir=None):
    data = []
    active = []
    ids = []
    next_id = 1
    # Map "KringName.nr" -> element id for sitplan references
    ref_to_id = {}
    current_kring_name = None
    kring_counter = 0

    def get_next_kring_name():
        nonlocal kring_counter
        # Generate A, B, C, ... Z, AA, AB, ...
        name = ""
        n = kring_counter
        while True:
            name = chr(ord('A') + n % 26) + name
            n = n // 26 - 1
            if n < 0:
                break
        kring_counter += 1
        return name

    def walk(node, parent=0, indent=0, auto_nr=None, kring_name=None):
        nonlocal next_id, current_kring_name
        my_id = next_id
        next_id += 1

        # Normalize shorthand syntax
        node = normalize_node(node)
        node_type = node["type"]
        # Apply defaults then overlay with provided props
        props = apply_defaults(node_type, node.get("props", {}))
        props["type"] = node_type

        # nr: explicit > auto-assigned > empty
        if "nr" in node:
            props = {"nr": str(node["nr"]), **props}
        elif auto_nr is not None:
            props = {"nr": str(auto_nr), **props}
        elif "nr" not in props:
            props = {"nr": "", **props}

        item = {
            "id": my_id,
            "parent": parent,
            "indent": indent,
            "collapsed": False,
            "props": props,
            "sourcelist": None
        }

        data.append(item)
        active.append(True)
        ids.append(my_id)

        # Track Kring name for sitplan refs
        if node_type == "Kring":
            current_kring_name = get_next_kring_name()

        # Register ref for sitplan (KringName.nr)
        if current_kring_name and props.get("nr"):
            ref = f"{current_kring_name}.{props['nr']}"
            ref_to_id[ref] = my_id

        # Auto-number children within Kring
        is_kring = node_type == "Kring"
        child_nr = 1
        for child in node.get("children", []):
            # Normalize child to check for explicit nr
            child_norm = normalize_node(child)
            if is_kring and "nr" not in child_norm:
                walk(child_norm, my_id, indent + 1, auto_nr=child_nr, kring_name=current_kring_name)
                child_nr += 1
            else:
                walk(child_norm, my_id, indent + 1, kring_name=current_kring_name)

    # Support single root or list of roots
    roots = tree if isinstance(tree, list) else [tree]
    for root in roots:
        walk(root)

    return {
        "currentView": "",
        "idToOrdinalMap": {},
        "length": len(data),
        "data": data,
        "active": active,
        "id": ids,
        "print_table": {
            "height": 562,
            "maxwidth": 484,
            "displaypage": 0,
            "enableAutopage": True,
            "printPageMode": "all",
            "printPageRange": "",
            "pages": [{"height": 562, "start": 0, "stop": 484, "info": "2 x 230V ~50 Hz"}],
            "pagemarkers": {"markers": []},
            "modevertical": "alles",
            "starty": 0,
            "stopy": 562,
            "papersize": "A4"
        },
        "properties": {
            "filename": "",
            "owner": (properties or {}).get("owner", "").strip().replace("\n", "<br>"),
            "installer": (properties or {}).get("installer", "").strip().replace("\n", "<br>"),
            "control": (properties or {}).get("control", "").strip().replace("\n", "<br>"),
            "info": (properties or {}).get("info", "2 x 230V ~50 Hz").strip().replace("\n", "<br>"),
            "currentView": None,
            "legacySchakelaars": False
        },
        "curid": next_id,
        "mode": "edit",
        "sitplan": None,
        "sitplanjson": build_sitplanjson(sitplan, ref_to_id, yaml_dir),
        "sitplanview": None
    }

def load_image_as_svg(image_path, size):
    """Load an image file and wrap it in SVG with base64 encoding."""
    import base64

    ext = image_path.lower().rsplit('.', 1)[-1]
    mime_types = {'png': 'image/png', 'jpg': 'image/jpeg', 'jpeg': 'image/jpeg', 'gif': 'image/gif'}
    mime_type = mime_types.get(ext, 'image/png')

    with open(image_path, 'rb') as f:
        img_data = base64.b64encode(f.read()).decode('ascii')

    width, height = size
    return f'<svg width="{width}" height="{height}"><image xlink:href="data:{mime_type};base64,{img_data}" width="{width}" height="{height}"/></svg>'

def build_sitplanjson(sitplan, ref_to_id, yaml_dir=None):
    """Build sitplanjson from YAML sitplan section."""
    if not sitplan:
        return {
            "numPages": 1,
            "activePage": 1,
            "defaults": {"fontsize": 11, "scale": 0.7, "rotate": 0},
            "elements": []
        }

    # Get defaults
    defaults = sitplan.get("defaults", {})
    default_fontsize = defaults.get("fontsize", 10)
    default_scale = defaults.get("scale", 0.25)
    default_rotate = defaults.get("rotate", 0)

    elements = []

    # Support both 'layers' (new format) and 'elements' (old format)
    layer_list = sitplan.get("layers", sitplan.get("elements", []))

    for layer in layer_list:
        pos = layer.get("pos", [0, 0])
        scale = layer.get("scale", default_scale)
        rotate = layer.get("rotate", default_rotate)
        page = layer.get("page", 1)
        color = layer.get("color", "#000000")

        if "image" in layer:
            # This is an image layer
            size = layer.get("size", [200, 200])
            image_file = layer["image"]

            # Load image from file
            svg_content = ""
            if yaml_dir:
                image_path = os.path.join(yaml_dir, image_file)
                if os.path.exists(image_path):
                    svg_content = load_image_as_svg(image_path, size)
                else:
                    print(f"Warning: Image file '{image_file}' not found")

            elements.append({
                "page": page,
                "posx": pos[0],
                "posy": pos[1],
                "sizex": size[0],
                "sizey": size[1],
                "labelposx": pos[0] + size[0] * 0.2,
                "labelposy": pos[1] + 1,
                "labelfontsize": default_fontsize,
                "adrestype": None,
                "adres": None,
                "adreslocation": "rechts",
                "rotate": rotate,
                "scale": scale,
                "movable": True,
                "color": color,
                "svg": svg_content,
                "electroItemId": None
            })
        else:
            # This is an electrical element reference
            ref = layer.get("ref")
            elem_id = layer.get("id")  # Support internal ID as alternative

            if ref:
                electro_id = ref_to_id.get(ref)
                if not electro_id:
                    print(f"Warning: sitplan ref '{ref}' not found in schema")
                    continue
            elif elem_id:
                electro_id = elem_id
            else:
                continue

            labelfontsize = layer.get("labelfontsize", default_fontsize)

            # Calculate size based on scale (default symbol size ~200x200)
            base_size = 200
            sizex = int(base_size * scale) + 1
            sizey = int(base_size * scale)

            # Label position offset
            labelposx = pos[0] + sizex * 0.2
            labelposy = pos[1] + 1

            elements.append({
                "page": page,
                "posx": pos[0],
                "posy": pos[1],
                "sizex": sizex,
                "sizey": sizey,
                "labelposx": labelposx,
                "labelposy": labelposy,
                "labelfontsize": labelfontsize,
                "adrestype": "auto",
                "adres": None,
                "adreslocation": layer.get("adreslocation", "rechts"),
                "rotate": rotate,
                "scale": scale,
                "movable": True,
                "color": color,
                "svg": "",
                "electroItemId": electro_id
            })

    return {
        "numPages": sitplan.get("numPages", 1),
        "activePage": sitplan.get("activePage", 1),
        "defaults": {
            "fontsize": default_fontsize,
            "scale": default_scale,
            "rotate": default_rotate
        },
        "elements": elements
    }

def yaml_to_eds(yaml_path, eds_path):
    with open(yaml_path, "r", encoding="utf-8") as f:
        tree = yaml.safe_load(f)

    # Get directory of YAML file for resolving relative image paths
    yaml_dir = os.path.dirname(os.path.abspath(yaml_path))

    # Extract properties and sitplan if present
    properties = None
    sitplan = None
    if isinstance(tree, dict):
        properties = tree.pop("properties", None)
        sitplan = tree.pop("sitplan", None)
        tree = tree.get("elements", [])

    json_to_eds(compile_tree(tree, properties, sitplan, yaml_dir), eds_path)
    print(f"Converted {yaml_path} -> {eds_path}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python convert.py input.yaml [output.eds]")
        print("       If output.eds is omitted, uses input name with .eds extension")
        sys.exit(1)

    yaml_path = sys.argv[1]
    if len(sys.argv) >= 3:
        eds_path = sys.argv[2]
    else:
        eds_path = yaml_path.rsplit(".", 1)[0] + ".eds"

    yaml_to_eds(yaml_path, eds_path)

File: tools-ts/src/convert.ts
/**
 * YAML to EDS converter
 *
 * Converts human-readable YAML format to EDS binary format
 */

import * as fs from 'fs';
import * as path from 'path';
import * as YAML from 'yaml';
import { jsonToEds, EdsData, EdsItem, SitplanJson, SitplanElement } from './eds-codec.js';
import { SCHEMA, getDefaults, getTypeNames } from './schema.js';

// Types for YAML input
interface YamlNode {
  type?: string;
  props?: Record<string, unknown>;
  children?: YamlNode[];
  nr?: number | string;
  [key: string]: unknown;
}

interface YamlProperties {
  owner?: string;
  installer?: string;
  control?: string;
  info?: string;
}

interface YamlSitplanLayer {
  pos: [number, number];
  ref?: string;
  id?: number;
  scale?: number;
  rotate?: number;
  page?: number;
  color?: string;
  image?: string;
  size?: [number, number];
  labelfontsize?: number;
  adreslocation?: string;
}

interface YamlSitplan {
  numPages?: number;
  activePage?: number;
  defaults?: {
    fontsize?: number;
    scale?: number;
    rotate?: number;
  };
  layers?: YamlSitplanLayer[];
  elements?: YamlSitplanLayer[]; // Old format
}

interface YamlMetadata {
  schema_version?: string;
  tool_version?: string;
  generated_at?: string;
}

interface YamlDocument {
  metadata?: YamlMetadata;
  properties?: YamlProperties;
  sitplan?: YamlSitplan;
  elements?: YamlNode[];
}

const TYPE_NAMES = getTypeNames();

/**
 * Convert shorthand syntax to standard format
 */
function normalizeNode(node: unknown): YamlNode {
  if (typeof node === 'string') {
    // Simple string: "Lichtcircuit"
    return { type: node };
  }

  if (typeof node === 'object' && node !== null) {
    const nodeObj = node as Record<string, unknown>;

    if ('type' in nodeObj) {
      // Already in standard format
      return nodeObj as YamlNode;
    }

    // Check if it's a type-name-as-key format
    for (const typeName of TYPE_NAMES) {
      if (typeName in nodeObj) {
        const value = (nodeObj[typeName] || {}) as Record<string, unknown>;
        const result: YamlNode = { type: typeName };

        if (typeof value === 'object' && value !== null) {
          // Extract children and props
          const { children, props, ...rest } = value as {
            children?: YamlNode[];
            props?: Record<string, unknown>;
            [key: string]: unknown;
          };

          // Combine explicit props with remaining keys
          const combinedProps = { ...(props || {}), ...rest };
          if (Object.keys(combinedProps).length > 0) {
            result.props = combinedProps;
          }
          if (children) {
            result.children = children;
          }
        }
        return result;
      }
    }
  }

  return node as YamlNode;
}

/**
 * Apply defaults for a node type, then overlay with provided props
 */
function applyDefaults(
  nodeType: string,
  props: Record<string, unknown>
): Record<string, unknown> {
  const defaults = getDefaults(nodeType);
  return { ...defaults, ...props };
}

/**
 * Generate next Kring name (A, B, C, ... Z, AA, AB, ...)
 */
function createKringNameGenerator(): () => string {
  let counter = 0;

  return () => {
    let name = '';
    let n = counter;
    while (true) {
      name = String.fromCharCode('A'.charCodeAt(0) + (n % 26)) + name;
      n = Math.floor(n / 26) - 1;
      if (n < 0) break;
    }
    counter++;
    return name;
  };
}

/**
 * Load image file and wrap in SVG with base64 encoding
 */
function loadImageAsSvg(imagePath: string, size: [number, number]): string {
  const ext = path.extname(imagePath).toLowerCase().slice(1);
  const mimeTypes: Record<string, string> = {
    png: 'image/png',
    jpg: 'image/jpeg',
    jpeg: 'image/jpeg',
    gif: 'image/gif',
  };
  const mimeType = mimeTypes[ext] || 'image/png';

  const imgData = fs.readFileSync(imagePath).toString('base64');
  const [width, height] = size;

  return `<svg width="${width}" height="${height}"><image xlink:href="data:${mimeType};base64,${imgData}" width="${width}" height="${height}"/></svg>`;
}

/**
 * Build sitplanjson from YAML sitplan section
 */
function buildSitplanJson(
  sitplan: YamlSitplan | undefined,
  refToId: Map<string, number>,
  yamlDir?: string
): SitplanJson {
  if (!sitplan) {
    return {
      numPages: 1,
      activePage: 1,
      defaults: { fontsize: 11, scale: 0.7, rotate: 0 },
      elements: [],
    };
  }

  const defaults = sitplan.defaults || {};
  const defaultFontsize = defaults.fontsize ?? 10;
  const defaultScale = defaults.scale ?? 0.25;
  const defaultRotate = defaults.rotate ?? 0;

  const elements: SitplanElement[] = [];

  // Support both 'layers' (new format) and 'elements' (old format)
  const layerList = sitplan.layers || sitplan.elements || [];

  for (const layer of layerList) {
    const pos = layer.pos || [0, 0];
    const scale = layer.scale ?? defaultScale;
    const rotate = layer.rotate ?? defaultRotate;
    const page = layer.page ?? 1;
    const color = layer.color ?? '#000000';

    if (layer.image) {
      // This is an image layer
      const size = layer.size || [200, 200];

      let svgContent = '';
      if (yamlDir) {
        const imagePath = path.join(yamlDir, layer.image);
        if (fs.existsSync(imagePath)) {
          svgContent = loadImageAsSvg(imagePath, size);
        } else {
          console.warn(`Warning: Image file '${layer.image}' not found`);
        }
      }

      elements.push({
        page,
        posx: pos[0],
        posy: pos[1],
        sizex: size[0],
        sizey: size[1],
        labelposx: pos[0] + size[0] * 0.2,
        labelposy: pos[1] + 1,
        labelfontsize: defaultFontsize,
        adrestype: null,
        adres: null,
        adreslocation: 'rechts',
        rotate,
        scale,
        movable: true,
        color,
        svg: svgContent,
        electroItemId: null,
      });
    } else {
      // This is an electrical element reference
      const ref = layer.ref;
      let electroId: number | undefined;

      if (ref) {
        electroId = refToId.get(ref);
        if (electroId === undefined) {
          console.warn(`Warning: sitplan ref '${ref}' not found in schema`);
          continue;
        }
      } else if (layer.id !== undefined) {
        electroId = layer.id;
      } else {
        continue;
      }

      const labelfontsize = layer.labelfontsize ?? defaultFontsize;

      // Calculate size based on scale
      const baseSize = 200;
      const sizex = Math.floor(baseSize * scale) + 1;
      const sizey = Math.floor(baseSize * scale);

      elements.push({
        page,
        posx: pos[0],
        posy: pos[1],
        sizex,
        sizey,
        labelposx: pos[0] + sizex * 0.2,
        labelposy: pos[1] + 1,
        labelfontsize,
        adrestype: 'auto',
        adres: null,
        adreslocation: layer.adreslocation ?? 'rechts',
        rotate,
        scale,
        movable: true,
        color,
        svg: '',
        electroItemId: electroId,
      });
    }
  }

  return {
    numPages: sitplan.numPages ?? 1,
    activePage: sitplan.activePage ?? 1,
    defaults: {
      fontsize: defaultFontsize,
      scale: defaultScale,
      rotate: defaultRotate,
    },
    elements,
  };
}

/**
 * Compile YAML tree to EDS data structure
 */
function compileTree(
  tree: YamlNode[],
  properties?: YamlProperties,
  sitplan?: YamlSitplan,
  yamlDir?: string
): EdsData {
  const data: EdsItem[] = [];
  const active: boolean[] = [];
  const ids: number[] = [];
  let nextId = 1;

  // Map "KringName.nr" -> element id for sitplan references
  const refToId = new Map<string, number>();
  const getNextKringName = createKringNameGenerator();
  let currentKringName: string | null = null;

  function walk(
    node: YamlNode,
    parent = 0,
    indent = 0,
    autoNr?: number
  ): void {
    const myId = nextId++;

    // Normalize shorthand syntax
    const normalizedNode = normalizeNode(node);
    const nodeType = normalizedNode.type!;

    // Apply defaults then overlay with provided props
    const props = applyDefaults(nodeType, normalizedNode.props || {});
    props.type = nodeType;

    // nr: explicit > auto-assigned > empty
    if ('nr' in normalizedNode) {
      props.nr = String(normalizedNode.nr);
    } else if (autoNr !== undefined) {
      props.nr = String(autoNr);
    } else if (!('nr' in props)) {
      props.nr = '';
    }

    const item: EdsItem = {
      id: myId,
      parent,
      indent,
      collapsed: false,
      props,
      sourcelist: null,
    };

    data.push(item);
    active.push(true);
    ids.push(myId);

    // Track Kring name for sitplan refs
    if (nodeType === 'Kring') {
      currentKringName = getNextKringName();
    }

    // Register ref for sitplan (KringName.nr)
    if (currentKringName && props.nr) {
      const ref = `${currentKringName}.${props.nr}`;
      refToId.set(ref, myId);
    }

    // Auto-number children within Kring
    const isKring = nodeType === 'Kring';
    let childNr = 1;

    for (const child of normalizedNode.children || []) {
      const childNorm = normalizeNode(child);
      if (isKring && !('nr' in childNorm)) {
        walk(childNorm, myId, indent + 1, childNr);
        childNr++;
      } else {
        walk(childNorm, myId, indent + 1);
      }
    }
  }

  // Process all root elements
  for (const root of tree) {
    walk(root);
  }

  return {
    currentView: '',
    idToOrdinalMap: {},
    length: data.length,
    data,
    active,
    id: ids,
    print_table: {
      height: 562,
      maxwidth: 484,
      displaypage: 0,
      enableAutopage: true,
      printPageMode: 'all',
      printPageRange: '',
      pages: [
        {
          height: 562,
          start: 0,
          stop: 484,
          info: properties?.info || '2 x 230V ~50 Hz',
        },
      ],
      pagemarkers: { markers: [] },
      modevertical: 'alles',
      starty: 0,
      stopy: 562,
      papersize: 'A4',
    },
    properties: {
      filename: '',
      owner: (properties?.owner || '').trim().replace(/\n/g, '<br>'),
      installer: (properties?.installer || '').trim().replace(/\n/g, '<br>'),
      control: (properties?.control || '').trim().replace(/\n/g, '<br>'),
      info: (properties?.info || '2 x 230V ~50 Hz').trim().replace(/\n/g, '<br>'),
      currentView: null,
      legacySchakelaars: false,
    },
    curid: nextId,
    mode: 'edit',
    sitplan: null,
    sitplanjson: buildSitplanJson(sitplan, refToId, yamlDir),
    sitplanview: null,
  };
}

/**
 * Convert YAML file to EDS file
 */
export function yamlToEds(yamlPath: string, edsPath: string): void {
  const yamlContent = fs.readFileSync(yamlPath, 'utf-8');
  const doc = YAML.parse(yamlContent) as YamlDocument;

  // Get directory of YAML file for resolving relative image paths
  const yamlDir = path.dirname(path.resolve(yamlPath));

  // Extract sections
  const properties = doc.properties;
  const sitplan = doc.sitplan;
  const elements = doc.elements || [];

  const edsData = compileTree(elements, properties, sitplan, yamlDir);
  jsonToEds(edsData, edsPath);

  console.log(`Converted ${yamlPath} -> ${edsPath}`);
}

export { compileTree, normalizeNode, applyDefaults };
